Philips.CodeAnalysis.MaintainabilityAnalyzers/Maintainability/PreferReadOnlyParametersAnalyzer.cs

```diff
@@ -0,0 +1,108 @@
+﻿// © 2023 Koninklijke Philips N.V. See License.md in the project root for license information.
+
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Philips.CodeAnalysis.Common;
+
+namespace Philips.CodeAnalysis.MaintainabilityAnalyzers.Maintainability
+{
+	[DiagnosticAnalyzer(LanguageNames.CSharp)]
+	public class PreferReadOnlyParametersAnalyzer : SingleDiagnosticAnalyzer<MethodDeclarationSyntax, PreferReadOnlyParametersNodeAction>
+	{
+		private const string Title = @"Prefer readonly parameters";
+		public const string MessageFormat = @"Make parameter '{0}' a readonly type";
+		private const string Description = @"Method parameters that are not written to, can be made readonly";
+		public PreferReadOnlyParametersAnalyzer()
+			: base(DiagnosticId.PreferReadOnlyParameters, Title, MessageFormat, Description, Categories.Maintainability)
+		{ }
+	}
+
+	public class PreferReadOnlyParametersNodeAction : SyntaxNodeAction<MethodDeclarationSyntax>
+	{
+		private static readonly List<string> ReadWriteCollections = new() { "System.Array", "System.Collections.Generic.List", "System.Span" };
+		private readonly NamespaceIgnoringComparer _comparer = new();
+
+		public override void Analyze()
+		{
+			var parameters = Node.ParameterList.Parameters;
+			if (!parameters.Any())
+			{
+				return;
+			}
+
+			var aliases = Helper.GetUsingAliases(Node);
+			var collections = parameters.Where(p => IsReadWriteCollection(p.Type, aliases));
+			if (!collections.Any())
+			{
+				return;
+			}
+			var accesses = Node.DescendantNodes().OfType<MemberAccessExpressionSyntax>().ToList();
+			var setters = Node.DescendantNodes()
+				.OfType<AssignmentExpressionSyntax>()
```

Comment:
DataFlowAnalysis could complement this Analyzer, as it can follow the local variables. However, it cannot replace its functionality completely as I understand it.

Reply:
ok. I was just thinking that if it worked, it would be more foolproof.

Reply:
I'll look into doing something like [Data Flow Analysis](https://github.com/dotnet/roslyn-analyzers/blob/main/docs/Writing%20dataflow%20analysis%20based%20analyzers.md)

Reply:
Ok, dug deep into DataFlowAnalysis and to be honest I drowned. Documentation is out-of-date, the example code doesn't compile and I can't get my scenario in there nicely.